File: week1/SegmentTree.java
/*
Question 4: Range Performance Monitor (Segment Tree)

Support:
1) update(index, value)
2) queryMax(left, right)

Both operations must run in O(log N).
*/

class SegmentTree {

    private int[] segmentTree;
    private int totalElements;

    public SegmentTree(int[] inputArray) {
        totalElements = inputArray.length;
        segmentTree = new int[4 * totalElements];
        buildTree(inputArray, 0, 0, totalElements - 1);
    }

    private void buildTree(
            int[] inputArray,
            int currentNode,
            int startIndex,
            int endIndex) {

        if (startIndex == endIndex) {
            segmentTree[currentNode] = inputArray[startIndex];
            return;
        }

        int middleIndex = (startIndex + endIndex) / 2;

        buildTree(inputArray, 2 * currentNode + 1,
                startIndex, middleIndex);

        buildTree(inputArray, 2 * currentNode + 2,
                middleIndex + 1, endIndex);

        segmentTree[currentNode] =
                Math.max(segmentTree[2 * currentNode + 1],
                        segmentTree[2 * currentNode + 2]);
    }

    public void updateValue(int indexToUpdate, int newValue) {
        updateTree(0, 0, totalElements - 1,
                indexToUpdate, newValue);
    }

    private void updateTree(
            int currentNode,
            int startIndex,
            int endIndex,
            int indexToUpdate,
            int newValue) {

        if (startIndex == endIndex) {
            segmentTree[currentNode] = newValue;
            return;
        }

        int middleIndex = (startIndex + endIndex) / 2;

        if (indexToUpdate <= middleIndex) {
            updateTree(2 * currentNode + 1,
                    startIndex, middleIndex,
                    indexToUpdate, newValue);
        } else {
            updateTree(2 * currentNode + 2,
                    middleIndex + 1, endIndex,
                    indexToUpdate, newValue);
        }

        segmentTree[currentNode] =
                Math.max(segmentTree[2 * currentNode + 1],
                        segmentTree[2 * currentNode + 2]);
    }

    public int queryMax(int leftRange, int rightRange) {
        return queryTree(0, 0, totalElements - 1,
                leftRange, rightRange);
    }

    private int queryTree(
            int currentNode,
            int startIndex,
            int endIndex,
            int leftRange,
            int rightRange) {

        if (rightRange < startIndex || leftRange > endIndex)
            return Integer.MIN_VALUE;

        if (leftRange <= startIndex && endIndex <= rightRange)
            return segmentTree[currentNode];

        int middleIndex = (startIndex + endIndex) / 2;

        return Math.max(
                queryTree(2 * currentNode + 1,
                        startIndex, middleIndex,
                        leftRange, rightRange),
                queryTree(2 * currentNode + 2,
                        middleIndex + 1, endIndex,
                        leftRange, rightRange)
        );
    }
}

File: week1/SegmentTree.cpp
/*
Question 4: Range Performance Monitor (Segment Tree)

Support:
1) update(index, value)
2) queryMax(left, right)

Both operations must run in O(log N).
*/

#include <bits/stdc++.h>
using namespace std;

class SegmentTree {
private:
    vector<int> tree;
    int n;

    void build(vector<int>& arr, int node, int start, int end) {
        if (start == end) {
            tree[node] = arr[start];
            return;
        }

        int mid = (start + end) / 2;

        build(arr, 2*node+1, start, mid);
        build(arr, 2*node+2, mid+1, end);

        tree[node] = max(tree[2*node+1], tree[2*node+2]);
    }

    void update(int node, int start, int end,
                int idx, int val) {

        if (start == end) {
            tree[node] = val;
            return;
        }

        int mid = (start + end) / 2;

        if (idx <= mid)
            update(2*node+1, start, mid, idx, val);
        else
            update(2*node+2, mid+1, end, idx, val);

        tree[node] = max(tree[2*node+1], tree[2*node+2]);
    }

    int query(int node, int start, int end,
              int l, int r) {

        if (r < start || l > end)
            return INT_MIN;

        if (l <= start && end <= r)
            return tree[node];

        int mid = (start + end) / 2;

        return max(
            query(2*node+1, start, mid, l, r),
            query(2*node+2, mid+1, end, l, r)
        );
    }

public:
    SegmentTree(vector<int>& arr) {
        n = arr.size();
        tree.resize(4*n);
        build(arr, 0, 0, n-1);
    }

    void updateValue(int idx, int val) {
        update(0, 0, n-1, idx, val);
    }

    int queryMax(int l, int r) {
        return query(0, 0, n-1, l, r);
    }
};
